File: ui/src/main/kotlin/listenwithrandos/app/ChatWindow.kt
package listenwithrandos.app

import kotlinx.html.ButtonType.submit
import kotlinx.html.InputType.text
import kotlinx.html.js.onChangeFunction
import kotlinx.html.js.onSubmitFunction
import listenwithrandos.domain.chats.Chat
import listenwithrandos.domain.chats.getChats
import listenwithrandos.domain.chats.sendChat
import listenwithrandos.state.AppAction
import listenwithrandos.state.SetChats
import listenwithrandos.state.State
import org.w3c.dom.HTMLInputElement
import react.*
import react.dom.*
import react.redux.rConnect
import redux.WrapperAction
import kotlin.js.Date

interface ChatWindowStateProps : RProps {
    var chats: List<Chat>
}

interface ChatWindowDispatchProps : RProps {
    var submitChat: (String) -> Unit
}

interface ChatWindowProps : ChatWindowStateProps, ChatWindowDispatchProps

interface ChatWindowState : RState {
    var newChat: String
}

class ChatWindow : RComponent<ChatWindowProps, ChatWindowState>() {
    override fun ChatWindowState.init() {
        newChat = ""
    }

    override fun RBuilder.render() {
        div {
            h1 { +"Chats" }

            props.chats.forEach { chat ->
                p { +"Me: ${chat.message}" }
            }
        }

        form {
            attrs.onSubmitFunction = { event ->
                event.preventDefault()

                props.submitChat(state.newChat)

                setState {
                    newChat = ""
                }
            }

            div(classes = "input-group") {
                input(type = text, classes = "form-control") {
                    attrs.value = state.newChat
                    attrs.onChangeFunction = { event ->
                        val target = event.target as HTMLInputElement

                        setState {
                            newChat = target.value
                        }
                    }
                }

                div(classes = "input-group-append") {
                    button(type = submit, classes = "btn btn-primary") {
                        +"Send"
                    }
                }
            }
        }
    }
}

fun sendAndGetChats(message: String, dispatch: (AppAction) -> WrapperAction) {
    sendChat(Chat(userId = "me", message = message, time = Date()))
        .then { getChats() }
        .then { dispatch(SetChats(it)) }
}

val connectedChatWindow = rConnect<State, AppAction, WrapperAction, RProps, ChatWindowStateProps, ChatWindowDispatchProps, ChatWindowProps>(
    mapStateToProps = { state, _ ->
        chats = state.chats
    },
    mapDispatchToProps = { dispatch, _ ->
        submitChat = { sendAndGetChats(it, dispatch) }
    }
)(ChatWindow::class.js.unsafeCast<RClass<RProps>>())

File: ui-new/src/app/ChatWindow/ChatWindow.js
// @flow

import * as React from 'react'
import {connect} from 'react-redux'
import type {State} from '../../domain/store/reducer/reducer'
import {SEND_CHAT} from '../../domain/store/actions/actions'
import type {AppAction, Chat} from '../../domain/store/actions/actions'
import type {Dispatch} from 'redux'

type ChatWindowStateProps = {
  chats: Array<Chat>
}

type ChatWindowDispatchProps = {
  sendChat(chat: Chat): void
}

type ChatWindowProps = ChatWindowStateProps & ChatWindowDispatchProps

type ChatWindowState = {
  newChat: string
}

class ChatWindow extends React.PureComponent<ChatWindowProps, ChatWindowState> {
  constructor() {
    super()

    this.state = {
      newChat: '',
    }
  }

  onInputChange = (event: SyntheticEvent<HTMLInputElement>) => {
    this.setState({
      newChat: event.currentTarget.value,
    })
  }

  onSubmit = (event: SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault()

    this.props.sendChat({
      userId: 'user',
      message: this.state.newChat,
      time: new Date(),
    })

    this.setState({newChat: ''})
  }

  render() {
    return (
      <React.Fragment>
        <div>
          <h1>Chats</h1>
          {
            this.props.chats.map(chat => (
              <p key={chat.id}>{chat.userId}: {chat.message}</p>
            ))
          }
        </div>

        <form onSubmit={this.onSubmit}>
          <div className="input-group">
            <input type="text"
                   className="form-control"
                   value={this.state.newChat}
                   onChange={this.onInputChange}
            />

            <div className="input-group-append">
              <button type="submit" className="btn btn-primary">Send</button>
            </div>
          </div>
        </form>
      </React.Fragment>
    )
  }
}

const mapStateToProps = (state: State) => ({
  chats: state.chats,
})

const mapDispatchToProps = (dispatch: Dispatch<AppAction>) => ({
  sendChat(chat) {
    dispatch({
      type: SEND_CHAT,
      chat,
    })
  },
})

export const ConnectedChatWindow = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChatWindow)
